server.c

```diff
@@ -2786,6 +2843,114 @@ add_tcp_handler(
+		return;
+
+	nsd->server_region = region_create(xalloc, free);
+	nsd->event_base = nsd_child_event_base();
+
+	nsd->next_zone_to_verify = zone;
+	nsd->verifier_count = 0;
+	nsd->verifier_limit = nsd->options->verifier_count;
+	size = sizeof(struct verifier) * nsd->verifier_limit;
+	nsd->verifiers = region_alloc_zero(nsd->server_region, size);
+
+	for(size_t i = 0; i < nsd->verifier_limit; i++) {
+		nsd->verifiers[i].nsd = nsd;
+		nsd->verifiers[i].zone = NULL;
+		nsd->verifiers[i].pid = -1;
+		nsd->verifiers[i].output_stream.fd = -1;
+		nsd->verifiers[i].output_stream.priority = LOG_INFO;
+		nsd->verifiers[i].error_stream.fd = -1;
+		nsd->verifiers[i].error_stream.priority = LOG_ERR;
+	}
+
+	event_set(&cmd_event, cmdsocket, EV_READ|EV_PERSIST, verify_handle_command, nsd);
+	if(event_base_set(nsd->event_base, &cmd_event) != 0 ||
+	   event_add(&cmd_event, NULL) != 0)
+	{
+		log_msg(LOG_ERR, "verify: could not add command event");
+		goto fail;
+	}
+
+	event_set(&sigchld_event, SIGCHLD, EV_SIGNAL|EV_PERSIST, verify_handle_exit, nsd);
+	if(event_base_set(nsd->event_base, &sigchld_event) != 0 ||
+	   event_add(&sigchld_event, NULL) != 0)
+	{
+		log_msg(LOG_ERR, "verify: could not add SIGCHLD event");
+		goto fail;
+	}
+
+	memset(msgs, 0, sizeof(msgs));
+	for (int i = 0; i < NUM_RECV_PER_SELECT; i++) {
+		queries[i] = query_create(nsd->server_region,
+			compressed_dname_offsets,
+			compression_table_size, compressed_dnames);
+		query_reset(queries[i], UDP_MAX_MESSAGE_LEN, 0);
+		iovecs[i].iov_base = buffer_begin(queries[i]->packet);
+		iovecs[i].iov_len = buffer_remaining(queries[i]->packet);
+		msgs[i].msg_hdr.msg_iov = &iovecs[i];
+		msgs[i].msg_hdr.msg_iovlen = 1;
+		msgs[i].msg_hdr.msg_name = &queries[i]->addr;
+		msgs[i].msg_hdr.msg_namelen = queries[i]->addrlen;
+	}
+
+	for (size_t i = 0; i < nsd->verify_ifs; i++) {
+		struct udp_handler_data *data;
+		data = region_alloc_zero(
+			nsd->server_region, sizeof(*data));
+		add_udp_handler(nsd, &nsd->verify_udp[i], data);
+	}
+
+	tcp_accept_handler_count = nsd->verify_ifs;
+	tcp_accept_handlers = region_alloc_array(nsd->server_region,
+		nsd->verify_ifs, sizeof(*tcp_accept_handlers));
+
+	for (size_t i = 0; i < nsd->verify_ifs; i++) {
+		struct tcp_accept_handler_data *data;
+		data = &tcp_accept_handlers[i];
+		memset(data, 0, sizeof(*data));
+		add_tcp_handler(nsd, &nsd->verify_tcp[i], data);
+	}
+
+	while(nsd->next_zone_to_verify != NULL &&
+	      nsd->verifier_count < nsd->verifier_limit)
+	{
+		verify_zone(nsd, nsd->next_zone_to_verify);
+		nsd->next_zone_to_verify = verify_next_zone(nsd, zone);
+	}
+
+	/* short-lived main loop */
+	event_base_dispatch(nsd->event_base);
+
+	/* remove command and exit event handlers */
+	event_del(&sigchld_event);
+	event_del(&cmd_event);
+
+fail:
+	event_base_free(nsd->event_base);
+	region_destroy(nsd->server_region);
+
+	nsd->event_base = NULL;
+	nsd->server_region = NULL;
+	assert(nsd->next_zone_to_verify == NULL);
+	assert(nsd->verifier_count == 0);
```

Comment:
The two assertions should be above the 'fail:' label.  They are not met when fail is used.  But they are met when a successful exit is performed.

Reply:
Indeed. Good catch.